pkg/container/docker_run.go

```diff
@@ -492,87 +492,122 @@ func (cr *containerReference) extractPath(env *map[string]string) common.Executo
+		return errors.WithStack(err)
+	}
+
+	resp, err := cr.cli.ContainerExecAttach(ctx, idResp.ID, types.ExecStartCheck{
+		Tty: cr.input.Tty,
+	})
+	if err != nil {
+		return errors.WithStack(err)
+	}
+	defer resp.Close()
+
+	var outWriter io.Writer
+	outWriter = cr.input.Stdout
+	if outWriter == nil {
+		outWriter = os.Stdout
+	}
+	errWriter := cr.input.Stderr
+	if errWriter == nil {
+		errWriter = os.Stderr
+	}
+
+	if cr.input.Tty {
+		// Avoid a freeze of programs expecting input like git asking for password and node
+		go func() {
+			c := 1
+			var err error
+			for c == 1 && err == nil {
+				c, err = resp.Conn.Write([]byte{4})
+				<-time.After(time.Second)
 			}
-		} else {
-			wd = cr.input.WorkingDir
-		}
-		logger.Debugf("Working directory '%s'", wd)
-
-		idResp, err := cr.cli.ContainerExecCreate(ctx, cr.id, types.ExecConfig{
-			User:         user,
-			Cmd:          cmd,
-			WorkingDir:   wd,
-			Env:          envList,
-			Tty:          isTerminal,
-			AttachStderr: true,
-			AttachStdout: true,
-		})
-		if err != nil {
-			return errors.WithStack(err)
-		}
+		}()
+	}
 
-		resp, err := cr.cli.ContainerExecAttach(ctx, idResp.ID, types.ExecStartCheck{
-			Tty: isTerminal,
-		})
-		if err != nil {
-			return errors.WithStack(err)
-		}
-		defer resp.Close()
+	if !cr.input.Tty || os.Getenv("NORAW") != "" {
+		_, err = stdcopy.StdCopy(outWriter, errWriter, resp.Reader)
+	} else {
+		_, err = io.Copy(outWriter, resp.Reader)
+	}
+	if err != nil {
+		logger.Error(err)
+	}
 
-		var outWriter io.Writer
-		outWriter = cr.input.Stdout
-		if outWriter == nil {
-			outWriter = os.Stdout
-		}
-		errWriter := cr.input.Stderr
-		if errWriter == nil {
-			errWriter = os.Stderr
-		}
+	inspectResp, err := cr.cli.ContainerExecInspect(ctx, idResp.ID)
+	if err != nil {
+		return errors.WithStack(err)
+	}
 
-		if !isTerminal || os.Getenv("NORAW") != "" {
-			_, err = stdcopy.StdCopy(outWriter, errWriter, resp.Reader)
-		} else {
-			_, err = io.Copy(outWriter, resp.Reader)
-		}
-		if err != nil {
-			logger.Error(err)
-		}
+	if inspectResp.ExitCode == 0 {
+		return nil
+	}
 
-		inspectResp, err := cr.cli.ContainerExecInspect(ctx, idResp.ID)
-		if err != nil {
-			return errors.WithStack(err)
-		}
+	return fmt.Errorf("exit with `FAILURE`: %v", inspectResp.ExitCode)
+}
 
-		if inspectResp.ExitCode == 0 {
-			return nil
+func (cr *containerReference) exec(cmd []string, env map[string]string, user, workdir string) common.Executor {
+	return func(ctx context.Context) error {
+		logger := common.Logger(ctx)
+		done := make(chan error)
+		go func() {
+			defer func() {
+				done <- errors.New("Invalid Operation")
+			}()
+			done <- cr.exec2(ctx, cmd, env, user, workdir)
+		}()
+		select {
+		case <-ctx.Done():
+			backgroundContext := context.Background()
+			err := cr.cli.ContainerKill(backgroundContext, cr.id, "kill")
+			if err != nil {
+				logger.Error(err)
+			}
+			err = cr.cli.ContainerStart(backgroundContext, cr.id, types.ContainerStartOptions{})
```

Comment:
I think restarting the container kind of defeats the purpose of this PR. If the intention is to allow to run post actions, shouldn't we rather check for a cancelled context in the step code, so that we can decide to run a `post` step there?